### PR TITLE
Add `Value` to eventually replace `ValueObject`

### PR DESCRIPTION
**Pull request by atifaziz**
This PR is a follow-up from PR #100 and supersedes it. It's less provocative 😉 in the sense that it makes the same improvements without breaking any compatibility. The public API remains the same since `ValueObject`, unlike in PR #100, isn't being removed anymore. Instead, this PR changes & replaces all _internal_ uses of `ValueObject` with the new _value type_ `Value`. The arguments for this are as follows (and remain pretty much the same as for PR #100):

- `ValueObject` is a `class` and causes an additional allocation per value. ~Converting it to a `struct` will require quite some review without much added benefits except saving on the wrapper allocation.~
- `Add` member of `ValueObject` was used in just one place and could be in-lined at the single call site (during leaf pattern matching); the latter was chosen. `Add` was the only reason `ValueObject` was read-write.
- Equality members like `Equals` and `GetHashCode` were never used or needed.
- ~It doesn't make the wrapped value any more strong-typed than using the bare `object`; instead it~ `ValueObject` does conversions in some cases that can fail, which isn't the case with `Value`.
- A `ValueObject` reference that can be null or an instance with a `null` for `Value` adds confusing semantics at best.
- Modern C# features like pattern matching are equally effective at succinctly encoding decisions based on type checks.

All tests still pass.

The new `Value` type is a read-only `struct` so it doesn't cause any allocations to wrap the underlying type. It's a discriminated union of the values it permits:

```c#
enum ValueKind { None, Boolean, Integer, String, StringList }
```

`StringList` is also a new read-only type that holds one or more strings laid out like a [_cons list_](https://en.wikipedia.org/wiki/Cons#Lists). It's stronger-typed than an `ArrayList` that `ValueObject` holds, but it's immutable nature and internal implementation layout also means that it causes a cell allocation per entry. While it's a list, its native and most efficient operations make it act more like a stack. In other words, you can only prepend (push/`cons`) items, remove the head item (pop/`cdr`) or request the item at the head (peek/`car`). `StringList` also supports _value equality_ by comparing its items for equality with those of another `StringList`.

While command/argument/option values are _internally_ maintained as `Value` instances now, they are converted to `ValueObject` at the very end before returning to the user code. This is done through a new overload of `Apply` that enables, what I'd like to call, _bring your own types_ (BYOT). That is, the new overload doesn't care about or enforce any representation for values as `ValueObject` or options as `IDictionary<string, ValueObject>`. The overload has the following _generic_ signature:

```c#
internal T Apply<T>(string doc, Tokens tokens,
                    IApplicationResultAccumulator<T> accumulator,
                    bool help = true, object version = null,
                    bool optionsFirst = false, bool exit = false)
```

There is a new abstraction in there called `IApplicationResultAccumulator<>` that's defined as:

```c#
interface IApplicationResultAccumulator<T>
{
    T New();
    T Command(T state, string name, bool value);
    T Command(T state, string name, int value);
    T Argument(T state, string name);
    T Argument(T state, string name, string value);
    T Argument(T state, string name, StringList value);
    T Option(T state, string name);
    T Option(T state, string name, bool value);
    T Option(T state, string name, string value);
    T Option(T state, string name, int value);
    T Option(T state, string name, StringList value);
    T Error(DocoptBaseException exception);
}
```

The parsed options and their values are fed through an implementation of the above accumulator interface that then accumulates/folds to some state type `T`. I have added an implementation (see `ValueObjectDictionaryAccumulator`) that accumulates commands, arguments and options as keyed `ValueObject` instances to `IDictionary<string, ValueObject>`, thus yielding full backwards compatibility. A second implementation (`ValueDictionaryAccumulator`) does that same but encodes values as `Value` instead of `ValueObject`. This second implementation is only used in tests in this PR, but the main motivation for its addition is for PR #77 (we can eventually think about exposing it publicly too). While it was not necessary, the tests were updated to use `Value` rather than `ValueObject` since the bulk cases were just initializing values rather than using any features from `ValueObject` and the code reads simpler now.

### CODE REVIEW

**Comment by atifaziz** [comment]
@smoothdeveloper Sorry it took so long to follow-up on your comments in PR #100, but my availability to work on this project fell considerably over the last 3 months. Nevertheless, I also took some time to understand `ValueObject` and all the valid instances of its use with commands, arguments and options. I believe that this PR addresses most of the concerns you raised as well as ideas you floated.

> `ValueObject` offers a bit of a statically typed API helping to reason about those things, while adhoc pattern matching makes it harder to reason about IMO.

The new `Value` type is a lot closer to `ValueObject` and fixes many of its problems that bothered from an allocation, conversions and correctness perspective.

> @atifaziz question: if you had Discriminated Unions available in C#, would you go with naked object rather than defining the finite set of possible cases this type ends up having at runtime currently?

`Value` is now a discriminated union. There is just no way to express them succinctly and in a closed fashion in C# today. Records comes close, but they are still classes and cause allocations. I believe value-records may be landing soon.

> …, I think a discriminated union, even hand desugared in C# solely, would still be better route for the long term maintenance than just `object`, no matter how expressive the pattern matching constructs are nowadays.

Agree and have thus have taken the approach to introduce `Value` in a more proper way. If I were to be totally honest, I think I was being somewhat lazy because I knew going the whole distance with a proper DU type would require a lot of time and effort and I was eager to return to work on PR #77 thinking that these _internal changes_ could always be improved on later.

File: src/DocoptNet/Docopt.cs [hunk 1]
```diff
@@ -25,6 +25,27 @@ partial class Docopt
...
+        {
+            return Apply(doc, tokens, ApplicationResultAccumulators.ValueObjectDictionary, help, version, optionsFirst, exit);
+        }
+
+        internal T Apply<T>(string doc, IApplicationResultAccumulator<T> accumulator)
+        {
+            return Apply(doc, new Tokens(Enumerable.Empty<string>(), typeof (DocoptInputErrorException)), accumulator);
+        }
+
+        internal T Apply<T>(string doc, ICollection<string> argv,
```
Comment (iwillspeak) [inline src/DocoptNet/Docopt.cs:37]: Do we want this to be public? I can't see how a user would inject an accumulator otherwise.

**Review by iwillspeak — APPROVED** [review]
I really like the direction that this is going. The accumulator interface looks super powerful. I've
knocked together an example of using it to aggregate into a class instance rather than a
dictionary over here: https://github.com/iwillspeak/docopt.net/commit/117cbe266a30560b7f653982ffce0279ff2b1a56

File: tests/DocoptNet.Tests/ApplicationResultAccumulator.cs [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+                var value = dict["<argument>"];
+                Assert.That(value, Is.InstanceOf<ValueObject>());
+                Assert.That(value.Value, Is.EqualTo("value"));
+            }
+
+            [Test]
+            public void Option_adds_entry_with_value_converted_to_object()
+            {
+                var dict = Accumulator.New();
+                dict = Accumulator.Argument(dict, "--option", "value");
```
[the remark this answers is not in the document]
Comment (atifaziz) [reply tests/DocoptNet.Tests/ApplicationResultAccumulator.cs]: Thanks for catching this!

File: src/DocoptNet/Docopt.cs (hunk 1 above)
Comment (atifaziz) [reply src/DocoptNet/Docopt.cs:37]: Yes, I think eventually. For now, I wanted to bring this as an internal design, refactoring and changes that can eventually be added to the public API.

**Comment by atifaziz** [comment]
> I really like the direction that this is going.

@iwillspeak Glad to hear and thanks for your thorough review!

> The accumulator interface looks super powerful.

Right, the idea was to put the user code in control and detach `Apply` from caring at all about the output or mandating it. Eventually, I think we can generalise it even further with a generator interface and layer the accumulator interface on top of it. I believe the generator version will be even simpler and more open/powerful. I might propose it as a separate PR.

> I've
> knocked together an example of using it to aggregate into a class instance rather than a
> dictionary over here: [iwillspeak@117cbe2](https://github.com/iwillspeak/docopt.net/commit/117cbe266a30560b7f653982ffce0279ff2b1a56)

Cool and really awesome to see another use case in action! I iterated [another version](https://github.com/atifaziz/docopt.net/commit/7119d9617e878a46f2f39c0905cae1edbba88181) where I removed some reflection, especially for [setting the value](https://github.com/iwillspeak/docopt.net/commit/117cbe266a30560b7f653982ffce0279ff2b1a56#diff-c9b38770f5f7df88244e52aad4515268f15f242db0ad88baa881fd9e948cbc2bR105). Instead, I went with quoted expressions, where one supplies a getter expression and an assignment/setter is dynamically inferred and compiled. One can then either provide the name-property mapping explicitly:

```c#
ApplicationResultAccumulators.Create<TestArguments>(
    ("command"                    , args => args.Command                   ),
    ("<argument>"                 , args => args.ArgArgument               ),
    ("--option"                   , args => args.FlagOption                ),
    ("--max-degree-of-parallelism", args => args.FlagMaxDegreeOfParallelism));
```

or just list the property and have the kebabised name inferred from the property name based on prefix rules (`Arg` for arguments, `Flag` for option and assume a command otherwise, which is what you had):

```c#
ApplicationResultAccumulators.Create<TestArguments>(
    args => args.Command,
    args => args.ArgArgument,
    args => args.FlagOption,
    args => args.FlagMaxDegreeOfParallelism);
```

**Comment by iwillspeak** [comment]
Yeah, all of those have their plus points. It seems like a nice abstraction.

**Review by voieducode — APPROVED** [review]

**Pull request merged**